pkgs/matcher/lib/src/iterable_matchers.dart

```diff
@@ -414,3 +414,61 @@ class _ContainsOnce extends _IterableMatcher {
           Description mismatchDescription, Map matchState, bool verbose) =>
       mismatchDescription.add(_test(item, matchState)!);
 }
+
+// Matches [Iterable]s which are sorted.
+Matcher isSorted<T>(Comparator<T> comparator) =>
+    _IsSorted<T, T>((t) => t, comparator);
+
+// Matches [Iterable]s which are sorted by the given key.
+Matcher isSortedBy<T, K extends Comparable<K>>(K Function(T) keyOf) =>
+    _IsSorted<T, K>(keyOf, (a, b) => a.compareTo(b));
+
+// Matches [Iterable]s which are sorted by the given key, using the given
+// comparator.
+Matcher isSortedByCompare<T, K>(
+        K Function(T) keyOf, Comparator<K> comparator) =>
+    _IsSorted(keyOf, comparator);
+
+class _IsSorted<T, K> extends _IterableMatcher<T> {
+  final K Function(T) _keyOf;
+  final Comparator<K> _comparator;
+
+  _IsSorted(K Function(T) keyOf, Comparator<K> comparator)
+      : _keyOf = keyOf,
+        _comparator = comparator;
+
+  String? _test(Iterable<T> item, Map matchState) {
+    if (item.length < 2) {
+      return null;
+    }
+
+    var prevElem = item.first;
+    var prevKey = _keyOf(prevElem);
+    for (final elem in item.skip(1)) {
+      var key = _keyOf(elem);
+      if (_comparator(prevKey, key) > 0) {
+        return StringDescription()
+            .add('found elements out of order ')
+            .add(prevElem.toString())
+            .add(' and ')
+            .add(elem.toString())
+            .toString();
```

Comment:
Seems like it can be:
```dart
    return StringDescription(
        'found elements out of order: $previousElement and $nextElement',
    ).toString();
```

Consider whether there should be some quote characters around the elements.

Reply:
Done.
Using `matchState` with `addDescriptionOf` for proper quoting.

This does mean that I couldn't use a string template here.